zapisy/apps/offer/plan/utils.py

```diff
@@ -0,0 +1,317 @@
+    teachers: Dict[str, List[TeacherInfo]]
+
+
+# Indexed by course name
+AssignmentsViewSummary = Dict[str, AssignmentsCourseInfo]
+
+
+# For get_votes function:
+class SingleYearVoteSummary(TypedDict):
+    # Total of points awarded in a vote.
+    total: int
+    # Number of voters who awarded this course with maximum number of votes.
+    count_max: int
+    # Number of voters that voted for this course
+    votes: int
+    # Number of enrolled students. None if course was not given that year.
+    enrolled: Optional[int]
+
+
+class ProposalVoteSummary(NamedTuple):
+    proposal: Proposal
+    semester: str
+    course_type: str
+    # Indexed by the academic year.
+    voting: Dict[str, SingleYearVoteSummary]
+
+
+# Indexed by the Proposal name
+VotingSummaryPerYear = Dict[str, ProposalVoteSummary]
+
+
+# For get_subjects_data:
+class SingleGroupData(TypedDict):
+    name: str
+    semester: str
+    teacher: str
+    teacher_code: int
+    # One of GROUP_TYPE_SHEETS
+    group_type: str
+    # Hours per week.
+    hours: int
+
+
+ProposalSummary = List[SingleGroupData]
+
+
+def propose(vote: ProposalVoteSummary):
+    """A simple, heuristic function to propose, whether the course should be taught in the upcoming year."""
+
+    current_year = SystemState.get_current_state().year
+    proposal = Proposal.objects.get(id=vote.proposal)
+    avg = SingleVote.objects.filter(state__year=current_year, value__gt=0).values('proposal').annotate(
+        total=Sum('value')).aggregate(Avg('total'))
+    previous_avg = 0
+    years = 0
+    percentage = 0.8
+
+    for year, values in vote.voting.items():
+        if not year == current_year:
+            years += 1
+            previous_avg = values['total']
+    if proposal.course_type.obligatory:
+        return True
+    if proposal.recommended_for_first_year:
+        return True
+    if vote.voting[current_year]['total'] >= avg['total__avg']:
+        return True
+    if years > 0 and vote.voting[current_year]['total'] >= percentage * (previous_avg / years):
+        return True
+    return False
+
+
+def get_subjects_data(subjects: List[Tuple[str, str, int]], years: List[str]) -> ProposalSummary:
+    """Function prepares data to create a csv.
+
+    Data returned by this function will be presented in a spreadsheet, where it will help with assigning classes. Given a course,
+    it will look for previous instances of the course (taking the newest one) and copy the group Assignments from it. If such
+    instance does not exist, it will assign only the owner of the course.
+
+    Args:
+        subjects: list of tuples (course name, course semester, course proposal). Each tuple represents single course.
+        years: list of years to look for data in.
+    """
+
+    course_data = {}
+
+    for subject in subjects:
+        course_name = subject[0]
+        course_semester = subject[1]
+        course_proposal = subject[2]
```

Comment:
Można nawet zwięźlej
```suggestion
    for course_name, course_semester, course_proposal in subjects:
```